sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/http/HttpDownloadFileTask.java

```diff
@@ -0,0 +1,203 @@
+            connection.connect();
+            String contentType = connection.getContentType();
+            if (contentType != null && contentType.contains("text")) {
+
+                /* This is not the expected APK file. Maybe the redirect could not be resolved. */
+                throw new IOException("The requested download does not appear to be a file.");
+            }
+
+            /* Download the release file. */
+            long totalBytesDownloaded = downloadFile(connection);
+            if (totalBytesDownloaded > 0) {
+                mDownloader.onDownloadComplete(mTargetFile);
+            }
+        } catch (IOException e) {
+            mDownloader.onDownloadError(e.getMessage());
+        } finally {
+            TrafficStats.clearThreadStatsTag();
+        }
+        return null;
+    }
+
+    /**
+     * Performs IO operation to download file through {@link URLConnection}.
+     * Saves the file to the {@link #mTargetFile).
+     *
+     * @param connection network connection,
+     * @return total number of downloaded bytes.
+     * @throws IOException if connection fails.
+     */
+    @SuppressWarnings("TryFinallyCanBeTryWithResources")
+    private long downloadFile(URLConnection connection) throws IOException {
+        InputStream input = null;
+        OutputStream output = null;
+        try {
+            input = new BufferedInputStream(connection.getInputStream());
+            output = new FileOutputStream(mTargetFile);
+            return copyStream(input, output, connection.getContentLength());
+        } finally {
+            try {
+                if (output != null) {
+                    output.close();
+                }
+                if (input != null) {
+                    input.close();
+                }
+            } catch (IOException ignored) {
+            }
+        }
+    }
+
+    /**
+     * Copies one stream into another and reports the progress.
+     *
+     * @param inputStream  the input stream.
+     * @param outputStream the output stream.
+     * @param lengthOfFile total number of bytes in the input stream. Used only for progress reporting.
+     * @return total number of processed bytes.
+     * @throws IOException if an I/O error occurs when reading or writing.
+     */
+    private long copyStream(@NonNull InputStream inputStream, @NonNull OutputStream outputStream, long lengthOfFile) throws IOException {
+        byte[] data = new byte[WRITE_BUFFER_SIZE];
+        int count;
+        long totalBytesDownloaded = 0, reported = 0;
+        while ((count = inputStream.read(data)) != -1) {
+            totalBytesDownloaded += count;
+            outputStream.write(data, 0, count);
+
+            /* Update the progress each UPDATE_PROGRESS_BYTES_COUNT bytes. */
+            if (totalBytesDownloaded >= reported + UPDATE_PROGRESS_BYTES_COUNT || totalBytesDownloaded == lengthOfFile) {
+                mDownloader.onDownloadProgress(totalBytesDownloaded, lengthOfFile);
```

Comment:
This is updated way too often:

```
09-26 18:08:15.181 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 128 KiB / 3525 KiB
09-26 18:08:15.241 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 256 KiB / 3525 KiB
09-26 18:08:15.271 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 384 KiB / 3525 KiB
09-26 18:08:15.311 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 512 KiB / 3525 KiB
09-26 18:08:15.341 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 640 KiB / 3525 KiB
09-26 18:08:15.361 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 768 KiB / 3525 KiB
09-26 18:08:15.401 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 896 KiB / 3525 KiB
09-26 18:08:15.441 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1024 KiB / 3525 KiB
09-26 18:08:15.481 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1152 KiB / 3525 KiB
09-26 18:08:15.521 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1280 KiB / 3525 KiB
09-26 18:08:15.551 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1408 KiB / 3525 KiB
09-26 18:08:15.571 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1536 KiB / 3525 KiB
09-26 18:08:15.601 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1664 KiB / 3525 KiB
09-26 18:08:15.631 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1792 KiB / 3525 KiB
09-26 18:08:15.661 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 1920 KiB / 3525 KiB
09-26 18:08:15.701 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2048 KiB / 3525 KiB
09-26 18:08:15.731 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2176 KiB / 3525 KiB
09-26 18:08:15.761 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2304 KiB / 3525 KiB
09-26 18:08:15.781 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2432 KiB / 3525 KiB
09-26 18:08:15.812 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2560 KiB / 3525 KiB
09-26 18:08:15.832 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2688 KiB / 3525 KiB
09-26 18:08:15.862 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2816 KiB / 3525 KiB
09-26 18:08:15.882 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 2944 KiB / 3525 KiB
09-26 18:08:15.902 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 3072 KiB / 3525 KiB
09-26 18:08:15.932 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 3200 KiB / 3525 KiB
09-26 18:08:15.962 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 3328 KiB / 3525 KiB
09-26 18:08:15.992 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 3456 KiB / 3525 KiB
09-26 18:08:16.012 5969-6040/com.microsoft.appcenter.sasquatch.project V/AppCenterDistribute: Downloading 2.3.1 (481) update: 3525 KiB / 3525 KiB
```

Reply:
Either consider time based or percentage based.

Reply:
I've added time based limit, but I have to mention that it should be often because it's used for displaying progress in UI